### DISCUSSION TITLE
Preserve color mapping across variables

### DISCUSSION THREAD

**Discussion by mcp292** [vega/altair#2993] [Q&A] [answered]
# Question
Is there a way to preserve color mapping across variables even when removing data?

# [MWE](https://altair-viz.github.io/gallery/simple_bar_chart.html)
You have a response that outshines all others, you want to show the scale, but then want to move in to a more zoomed in view to better compare the rest. In the example below, it would be less disorienting to have variable 'D' always be light blue.

```py
import altair as alt
import pandas as pd


source = pd.DataFrame({
    "variable": ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
    "response": [28, 55, 2043, 91, 81, 53, 19, 87, 52]
})

alt.Chart(source).mark_bar().encode(
    x="variable:N",
    y="response:Q",
    color="variable:N"
).save("color.png", webdriver="firefox")
```

![all](https://user-images.githubusercontent.com/49123398/227657563-9dfadbc9-58cc-451f-9a9a-a76a6579bf9b.png)

After filtering out variable C to get a better view:
```py
source = source[source["variable"] != 'C']
```

![filtered](https://user-images.githubusercontent.com/49123398/227657619-cb2a35c8-4da9-4ae8-a282-701c9fcce38d.png)

After masking variable C (I thought this would reserve a spot for it in the mapping but does the same as above):
```py
source = source.mask(source["variable"] == 'C')
```

![masked](https://user-images.githubusercontent.com/49123398/227657900-201cca25-d415-4f3b-9cd1-80a10afe1eea.png)

**Comment by joelostblom** [comment, marked as answer]
You can set the color encodings scale domain to a list of the values to include. On my phone right now so it's hard to type, but something like `alt.Scale(domain=[A, B, C... ]) ` inside the color encoding

**Reply by mcp292** [reply]
Outstanding! This works:
```py
import altair as alt
import pandas as pd


source = pd.DataFrame({
    "variable": ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
    "response": [28, 55, 2043, 91, 81, 53, 19, 87, 52]
})

color_domain = source["variable"].unique()

# source = source[source["variable"] != 'C']
source = source.mask(source["variable"] == 'C')

alt.Chart(source).mark_bar().encode(
    x="variable:N",
    y="response:Q",
    color=alt.Color(
        "variable:N",
        # scale=alt.Scale(domain=['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I']))
        scale=alt.Scale(domain=color_domain))
).save("color.png", webdriver="firefox")
```

**Reply by mcp292** [reply]
![all](https://user-images.githubusercontent.com/49123398/227657563-9dfadbc9-58cc-451f-9a9a-a76a6579bf9b.png) ![filtered with original color mapping](https://user-images.githubusercontent.com/49123398/227728528-50b4ac79-cf06-4044-b4f0-a5354cac857b.png)

**Reply by mcp292** [reply]
**NOTE:** [`unique()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html) returns items in order of appearance, altair sorts. To keep everything consistent you're gonna want to sort:
```py
color_domain = sorted(source["variable"].unique())
```
OR
```py
color_domain = source["variable"].unique()
color_domain.sort()  # numpy `sort()` sorts in place and returns `None`
```
